### PR TITLE
Allow Ivarators to take advantage of document range when possible.

### PR DESCRIPTION
**Pull request by apmoriarty**
- parse out datatype and uid when handed a document range
- List ivarator builds bounding ranges with datatype and uid

Note: Filter, Range, and Regex ivarators not modified to take advantage of the doc range.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java [hunk 1]
```diff
@@ -262,6 +263,17 @@ public B withIvaratorSourcePool(GenericObjectPool<SortedKeyValueIterator<Key,Val
     // a time filter
     private final TimeFilter timeFilter;
     
+    // construct selective bounding ranges if handed a document range
+    protected String uid;
+    protected String datatype;
+    // support for determining if we are in a high cardinality field value
+    protected long consecutiveFieldValueHits = 0L;
+    protected long consecutiveHitsThreshold = 40L; // based on intra-index block cost of a seek vs next
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:271]: should this be configurable?

**Review by FineAndDandy — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java (hunk 1 above)
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:271]: alternatively creating a ticket to make it configurable in the future would be fine

**Review by FineAndDandy — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:271]: Captured in https://github.com/NationalSecurityAgency/datawave/issues/1205

**Review by d-hwang — APPROVED** [review]

**Review by FineAndDandy — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java [hunk 2]
```diff
@@ -1066,6 +1102,44 @@ protected boolean addKey(Key topFiKey, Value value) throws IOException {
...
+        int fiUidIndex = nullIndex + 1;
+        
+        int fiUidLength = bytes.length() - fiUidIndex;
+        if (fiUidLength < uid.length()) {
+            // 'tis difficult to match a uid that is shorter than the target
+            return false;
+        }
+        
+        for (int i = 0; i < uid.length(); i++) {
+            if (uid.charAt(i) != bytes.byteAt(fiUidIndex)) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:1134]: Can we not compare the uid to the bytes the first time we loop through?

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java [hunk 3]
```diff
@@ -1170,6 +1244,73 @@ protected boolean addKey(Key topFiKey, Value value) throws IOException {
...
+                            // new value found, reset count
+                            consecutiveFieldValueHits = 0L;
+                            lastValue = fiKey.getValue();
+                        }
+                        
+                        if (lastCF.equals(top.getColumnFamily()) && lastValue.equals(fiKey.getValue())) {
+                            consecutiveFieldValueHits++;
+                        }
+                        
+                        if (consecutiveFieldValueHits > consecutiveHitsThreshold) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:1273]: Do we really want to use a consecutiveHitsThreshold?  Can't we simply seek each time to the next viable FI key given we have a value and datatype?

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java [hunk 4]
```diff
@@ -580,8 +592,27 @@ public void seek(Range r, Collection<ByteSequence> columnFamilies, boolean inclu
                 // seek our underlying source to the start of the incoming range
                 // expand the range as the underlying table may not actually contain the keys in this range as we are only returning keys
                 // as specified by the returnKeyType
-                Range seekRange = new Range(lastRangeSeeked.getStartKey(), lastRangeSeeked.isStartKeyInclusive(), (lastRangeSeeked.getEndKey() == null ? null
-                                : new Key(lastRangeSeeked.getEndKey().getRow()).followingKey(PartialKey.ROW)), false);
+                
+                // Parse datatype\0uid from seek range, if they exist
+                String lastCF = lastRangeSeeked.getStartKey().getColumnFamily().toString();
+                int index = lastCF.indexOf('\u0000');
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexCachingIteratorJexl.java:598]: Why not use the FiKey you developed below?

**Comment by hlgp** [comment]
Old MR with outstanding questions.  *nudge*

**Pull request closed without merging**